Optimistic deadness analysis

trentlo opened the pull request:
The optimistic deadness analysis can complement the original pessimistic
analysis and reduce false positives due to loop structures such as in
dynamic RNN.

In addition, a new rule {S,&,X} & ~X => S is added to try to cancel out
the leftover symbols after while loops. Otherwise, symbols keep accumulating
in the case of cascaded while loops.

trentlo committed:
Implement optimistic deadness analysis.

The optimistic deadness analysis can complement the original pessimistic
analysis and reduce false positives due to loop structures such as in
dynamic rnn.

In addition, a new rule {S,&,X} & ~X => S is added to try to cancel out
the leftover symbols after while loops. Otherwise, symbols keep accumulating
in the case of cascaded while loops.

trentlo commented:
Sanjoy, please help to take a look when you have a moment. Thanks.

@sanjoy @jlebar

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+        // make sure all nodes in the currently processing frame are visited
+        // before starting processing other frames.
+        string frame_name = cf_infos[staging_exit_vec.front()->id()].frame_name;
+        CHECK(staging_exit_vec.size() == num_exits[frame_name]);
```
Use `CHECK_EQ` because that will print out the two values if they're not equal.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1069,6 +1240,52 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+    bool is_success = false;
+    if (!force_pessimistic && !frame_name.empty()) {
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_tpo, /*use_optimistic_mode*/ true, &is_success));
```
Google style is `/*use_optimistic_mode=*/true` (internally we have a linter that will recognize this format and  complain if `use_optimistic_mode` does not match the argument name).

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1069,6 +1240,52 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+                                      graph_.source_node(), &i_frame_name));
+      TF_RETURN_IF_ERROR(GetRootFrame(tpo[i + 1], control_flow_info_,
+                                      graph_.source_node(), &next_frame_name));
+      if (i_frame_name == next_frame_name) {
```
This is a bit oddly written, how about restructuring the loop as:

```
frame_start = ...;
while (frame_start != tpo.size()) {
  for (i = frame_start + 1; ...) {
    // Find the first i such that frame_of[tpo[i]] != frame_of[tpo[frame_start]]
  }
  sub_tpo = ...
  frame_start = i + 1;
}
```

trentlo replied:
Will do.

trentlo replied:
Done.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1069,6 +1240,52 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+    }
+
+    string frame_name = control_flow_info_[tpo[i]->id()].frame_name;
+    absl::Span<Node*> sub_tpo(tpo.data() + frame_start, i - frame_start + 1);
```
Please add a `/*length=*/i - frame_start + 1` to make it obvious what the second argument is.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1069,6 +1240,52 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_tpo, /*use_optimistic_mode*/ false, nullptr));
+    }
+    if (VLOG_IS_ON(2)) {
```
This is not necessary -- `VLOG(2)` automatically checks this.  And it has some C++ macro cleverness to avoid evaluating the expressions being streamed into it so checking `VLOG_IS_ON(2)` in this case has no performance benefit either.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -937,8 +1092,21 @@ Status DeadnessAnalysisImpl::HandleMerge(Node* n,
+        // does not affect the result after pattern-matching into the
+        // AndRecurrence form.
+        string frame_name = control_flow_info_[n->id()].frame_name;
+        auto ret = frame_to_merge_node_.insert({frame_name, n});
```
Maybe s/`ret`/`insert_result`/ ?  Or you could avoid the temporary object altogether and do

`Node* representative = frame_to_merge_node_.insert({frame_name, n}).first->second;`

sanjoy reviewed (changes requested):
Thanks for the PR!

I have not yet been through the whole thing but I wanted to send out some early comments (mostly minor stylistic issues).

trentlo committed:
Address review comments.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -636,6 +642,36 @@ Predicate* PredicateFactory::MakeAndOrImpl(
+    std::vector<Predicate*> to_add;
+    for (Predicate* op : simplified_ops) {
+      if (op->kind() == Predicate::Kind::kAndRecurrence) {
+        AndRecurrencePredicate* and_rec =
```
Agreed.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -636,6 +642,36 @@ Predicate* PredicateFactory::MakeAndOrImpl(
+    std::vector<Predicate*> to_add;
+    for (Predicate* op : simplified_ops) {
+      if (op->kind() == Predicate::Kind::kAndRecurrence) {
+        AndRecurrencePredicate* and_rec =
```
If you refer to the location of the "*", it appears that in this file asterisk is with the type instead of the symbol name?

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -636,6 +642,36 @@ Predicate* PredicateFactory::MakeAndOrImpl(
+      if (op->kind() == Predicate::Kind::kAndRecurrence) {
+        AndRecurrencePredicate* and_rec =
+            static_cast<AndRecurrencePredicate*>(op);
+        if (negated_ops.count(and_rec->step())) {
```
Thx.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1100,11 +1318,11 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
     // The nodes added to should_revisit in the previous loop need to be
     // revisited now.  Reprocesing these initial nodes may add *their* consumers
     // to should_revisit, and these newly added nodes will also be processed by
-    // this very same loop.  Since we're traversing the graph in reverse post
```
I checked the wiki page: https://en.wikipedia.org/wiki/Topological_sorting

My read is that although rpo is (certainly) a valid topological order, it is just a kind of topological order. In other words, there are many kinds of topological orders. So, it would be more precise to say it is a topological order instead of rpo.

I added some more comments before the procedure and give a ref to the wiki page. Please let me know what you think.

sanjoy replied:
I see.  On second thought I agree that it is better to call this a topological sort since we're no longer doing a DFS to compute the order.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
 }
+
+Status GetRootFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+                    const Node* src_node, string* frame) {
```
Thanks for the suggestion.

trentlo replied:
Turns out we cannot in this case. I _guess_ if string_view is pointed to some entries in a absl::Span, it creates some unexpected behaviors when the Span is gone?

sanjoy replied:
The `Span` itself is a non-owning pointer.  I think the `string_view` can be safely used as long as the  vector of `ControlFlowInfo` stays live.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+  absl::flat_hash_map<string, size_t> num_exits;
+  std::vector<size_t> num_ready_inputs(g->num_node_ids(), 0);
+  Node* src_node = g->source_node();
+  std::deque<Node*> ready;
```
Thanks for the catch.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+    }
+  }
+
+  absl::flat_hash_map<string, std::vector<Node*>> staging_enter_vecs;
```
Let me think about it.

trentlo replied:
Your suggestion makes sense. Renamed to `read_enters_per_frame` and `ready_exits`.

trentlo commented:
Sanjoy, thanks for the review and early feedback. I addressed some of the review comments. Will continue when I have a moment.

trentlo committed:
Revise according to review comments.

Rename staging_enter_vecs and staging_exit_vec to ready_enters_per_frame
and ready_exits, respectively. Also, revise comments accordingly.

trentlo committed:
Address review comments.

Restructure the loop structures in Populate() of deadness analysis.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+Status GetFrameBasedTopologicalOrder(const Graph* g,
+                                     absl::Span<const ControlFlowInfo> cf_infos,
+                                     std::vector<Node*>* order) {
+  absl::flat_hash_map<string, size_t> num_enters;
```
Good suggestion.

sanjoy reviewed (changes requested):
Just over-communicate:  I'm currently waiting on you to address the first round of comments.  Once those are addressed I'll do a second review.  Please let me know if you'd prefer otherwise (that I review the code in the state it is in currently while you concurrently address the previous round of comments).

trentlo commented:
> Just over-communicate: I'm currently waiting on you to address the first round of comments. Once those are addressed I'll do a second review. Please let me know if you'd prefer otherwise (that I review the code in the state it is in currently while you concurrently address the previous round of comments).

Thanks for the status check. I pushed some more commits today. I think I should have addressed all the previous issues except the one related to using string_view to get the returned frame name in GetRootFrame. I will need to study this as I somehow encountered some unexpected behaviors when using string_view in this case.

However, please proceed your review when you are free. And let me know if all of your previous comments are addressed from your perspective.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
   return Status::OK();
 }
+
+Status GetRootFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
```
Comments added.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+    n = cfi_iter->parent_frame;
+    cfi_iter = &cfi_infos[n->id()];
+
+    if (depth++ > 5000) {
```
I prefer the old way as it is also used in `GetFrameName` and it is more informative.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,125 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+                                     std::vector<Node*>* order) {
+  absl::flat_hash_map<string, size_t> num_enters;
+  absl::flat_hash_map<string, size_t> num_exits;
+  std::vector<size_t> num_ready_inputs(g->num_node_ids(), 0);
```
Done.

trentlo reviewed (commented):
Just to make the communication more explicit by replying to all of your previous comments.

trentlo committed:
Use string_view to get frame names from GetRootFrame().

trentlo commented:
> > Just over-communicate: I'm currently waiting on you to address the first round of comments. Once those are addressed I'll do a second review. Please let me know if you'd prefer otherwise (that I review the code in the state it is in currently while you concurrently address the previous round of comments).
> 
> Thanks for the status check. I pushed some more commits today. I think I should have addressed all the previous issues except the one related to using string_view to get the returned frame name in `GetRootFrame()`. I will need to study this as I somehow encountered some unexpected behaviors when using string_view in this case.
> 
> Please proceed your review when you are free. And let me know if all of your previous comments are addressed from your perspective.

Sanjoy, you are right that we definitely can use string_view to get the frame name from GetFrameName(). I just was confused by myself. Never mind my previous comments about this.

Now, all of your previous comments should all have been addressed. Thanks for your time for the review again.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+  absl::flat_hash_map<absl::string_view, size_t> num_exits;
+  std::vector<size_t> num_ready_inputs(g->num_node_ids(), 0);
+  Node* src_node = g->source_node();
+  for (const auto* node : g->op_nodes()) {
```
Might be better to extract out the first half of this loop into a `GetEntryAndExitsPerRootFrame` helper.

trentlo replied:
I don't see a clear way to do that. Currently, the loop counts 3 things (`num_enters`, `num_exits`, and `num_ready_inputs`.) in one traversal. Not easy to factor out part of it.

sanjoy replied:
I was thinking we could do two traversals (more expensive but the code is clearer), but this is fine for now.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+      if (num_ready_inputs[out->id()] != out->in_edges().size()) continue;
+
+      bool is_root_level = cf_infos[out_id].parent_frame == src_node;
+      string frame_name = cf_infos[out_id].frame_name;
```
Can be an `absl::string_view`.

trentlo replied:
Will pay attention next time. Thx.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+
+      bool is_root_level = cf_infos[out_id].parent_frame == src_node;
+      string frame_name = cf_infos[out_id].frame_name;
+      if (IsEnter(out) && is_root_level) {
```
How about extracting out `IsEnter(out) && is_root_level` into a `IsRootEnter`?  Similarly for `IsRootExit`?

trentlo replied:
I don't see a good way to do that. If we make `IsRootExit()` a function, we need to pass `control_flow_info_` to it. Wouldn't that be a bit cumbersome?

sanjoy replied:
These would have to be member functions on `DeadnessAnalysisImpl` to avoid having to pass around `control_flow_info_` (which I agree would be cumbersome).

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+// many inputs of each node are ready; a node is ready to be scheduled if all
+// of its inputs are ready.
+// Ref. to https://en.wikipedia.org/wiki/Topological_sorting for details.
+Status GetFrameBasedTopologicalOrder(const Graph* g,
```
This should probably be a function on `DeadnessAnalysisImpl` so that we don't have to pass around `g` and `cf_infos`.

trentlo replied:
Make sense. Done.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+        // processing ready_enters_per_frame to make sure all nodes in the
+        // currently processing frame are visited before starting processing
+        // other frames.
+        string frame_name = cf_infos[ready_exits.front()->id()].frame_name;
```
Can be an `absl::string_view`.

trentlo replied:
Will pay attention next time. Thx.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+    }
+  }
+
+  CHECK(ready_enters_per_frame.empty() && ready_exits.empty());
```
A note on `CHECK`s -- if these checks can fail because of a malformed TF graphdef (e.g. bogus `Enter` nodes) then it we need to return a failed `Status` (perhaps an `errors::InvalidArgument`), not `CHECK` fail.  `CHECK`s are okay for internal invariants that have to be always true.

We don't follow this constraint everywhere, but we have some internal linting tools that tries to enforce this rule on new code.

trentlo replied:
Make it a `errors::InvalidArgument`. Will pay attention to the difference next time. Thx.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -937,8 +1102,21 @@ Status DeadnessAnalysisImpl::HandleMerge(Node* n,
+        // frame as the representative Merge node.  It is just convenient and
+        // does not affect the result after pattern-matching into the
+        // AndRecurrence form.
+        string frame_name = control_flow_info_[n->id()].frame_name;
```
`absl::string_view`.

trentlo replied:
Will pay attention next time. Thx.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1114,6 +1343,69 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+        continue;
+      }
+      Node* merge = n;
+      string frame_name = control_flow_info_[merge->id()].frame_name;
```
`absl::string_view`.

trentlo replied:
Will pay attention to it next time. Thx.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1114,6 +1343,69 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+      if (!insert_result.second) {
+        // If we have already seen this frame name, verify the predicate is the
+        // same as the previously seen one's.
+        AndRecurrencePredicate* curr_andrec =
```
You don't really need the cast here, right?  You could compare the `Predicate *` directly and also call `Predicate::ToString` directly.

trentlo replied:
Removed.

I used to have some old codes needing the casts. Oversight. Thx for the catch.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1114,6 +1343,69 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+  // all the predicates of the merge nodes in the same frame are the same.  If
+  // yes, report success.  If not, report failure and clear the assigned
+  // predicates.
+  if (use_optimistic_mode) {
```
I think this will be slightly more robust / obviously correct if in `HandleMerge` we keep track of the nodes we assigned optimistic predicates to (in some `std::vector<Node*>` field in the class) and check that those specific nodes converged to the same predicate here.

Also, maybe we can move this to a helper function?

trentlo replied:
I'd argue that the robustness is the same but the current implementation is more clear, because we explicitly check all the merges are the same for each frame. Note that in the case of nested while loops, each while can have different predicates while merges on the same level of while must have the same predicate (I added some comments to make this more explicit.) This can complicate the track and compare approach.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1170,22 +1462,14 @@ DeadnessAnalysisImpl::PredicateMapAsString() const {
+                         bool force_pessimistic) {
   DeadnessAnalysisImpl impl(&graph);
-  TF_RETURN_IF_ERROR(impl.Populate());
+  TF_RETURN_IF_ERROR(impl.Populate(force_pessimistic));
```
Is there value in testing with `force_pessimistic=true`?  I'd venture that we should test the default mode of analysis unless there is reason to do otherwise.

trentlo replied:
The unittest tests both cases if the results are different. On the other hand, the optimistic mode is tested when fallback is expected.

sanjoy commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -1069,6 +1250,53 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
+}
+
+Status DeadnessAnalysisImpl::PopulateFrame(absl::Span<Node* const> topo,
+                                           bool use_optimistic_mode,
```
Let's `CHECK` that `success` is not null if and only if `use_optimistic_mode` is true.  Alternately (I think this is slightly better) if the code guarded under `use_optimistic_mode` is extracted out into a function then we can have the caller do the "was optimism successful" check, letting us remove the `bool *success` out parameter altogether.

trentlo replied:
I chose the former suggestion, i.e., adding a `CHECK`.

I don't see a clear way to implement the later suggestion, as `success` is needed by `Populate()`. I thought about another way to remove `success` by moving the logic about running optimistic and then pessimistic modes from `Populate()` into `PopulateFrame()`. But this makes the `PopulateFrame` more complicated. So, I chose to just implement the `CHECK`.

sanjoy replied:
It is probably easier to show what I had in mind (if it works!) by sending a subsequent PR after this is checked in, so I'll do that. :)

trentlo replied:
Sounds fair. :)

sanjoy reviewed (changes requested):
Looks great, mostly some minor comments.

trentlo committed:
Address review comments.

- Make GetFrameBasedTopologicalOrder a member function of DeadnessAnalysisImpl.
- Use string_view when appropriate.
- Change force_pessimistic to enable_optimistic. And make it a non-default
  parameter.
- Rename num_enters/num_exits to num_enters_for_frame/num_exits_for_frame
  in GetFrameBasedTopologicalOrder().

trentlo committed:
Address review comments.

- Add some more checks and comments.
- Remove redundant casts.

trentlo committed:
Convert one more string to string_view.

trentlo committed:
Polish comments.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -636,6 +643,35 @@ Predicate* PredicateFactory::MakeAndOrImpl(
+    }
+    auto it = simplified_ops.begin();
+    while (it != simplified_ops.end()) {
+      if (to_remove.contains(*it)) {
```
I don't get this. `simplified_ops` seen in `simplified_ops`?

trentlo replied:
If you meant to check the op seen in `negation_ops` has the correspondents in `simplified_ops`, it should not be necessary because it is logically correct the op is not in `simplified_ops` (after it is cancelled out.)

sanjoy replied:
Argh, sorry for the weird comment.  I meant we should `DCHECK` that all elements in `to_remove` were seen in `simplified_ops`.

sanjoy replied:
I'll send in a PR once this is checked in to implement this.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -699,8 +735,9 @@ class DeadnessAnalysisImpl : public DeadnessAnalysis {
 
-  Status Populate();
-  Status PopulateWithReversePostOrder(absl::Span<Node* const> rpo);
+  Status Populate(bool force_pessimistic = false);
```
Make sense. Done.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+Status GetFrameBasedTopologicalOrder(const Graph* g,
+                                     absl::Span<const ControlFlowInfo> cf_infos,
+                                     std::vector<Node*>* order) {
+  absl::flat_hash_map<absl::string_view, size_t> num_enters;
```
Agreed. Done.

trentlo commented on tensorflow/compiler/jit/deadness_analysis.cc:
```diff
@@ -914,10 +954,135 @@ Status GetFullFrame(const Node* n, absl::Span<const ControlFlowInfo> cfi_infos,
+    }
+  }
+
+  std::deque<Node*> ready;
```
`dequeue` is needed. Comments added.

trentlo reviewed (commented):
Thx for the new round of review again. Pushed some more commits to address the review comments. Please help to take a look again when you have time.

trentlo committed:
Minor comment polishing.

sanjoy reviewed (approved):
LGTM!

trentlo commented:
Just to bug you one more time for this PR. Could you please help to check if the CI fails because of this PR?

sanjoy commented:
> Just to bug you one more time for this PR. Could you please help to check if the CI fails because of this PR?

Looks like the Windows CI is failing because of Eigen: https://source.cloud.google.com/results/invocations/f7e0c161-7c6f-4dd8-9bd5-8353a0e5acb8/targets/%2F%2Ftensorflow%2Ftools%2Fci_build%2Fbuilds:gen_win_out/log

But I already approved the CL internally so it looks like that failure did not block anything.  @rthadur is that correct?

rthadur commented:
Yes , it did not block
